Userland/Applications/Presenter/PresenterWidget.cpp

```diff
@@ -0,0 +1,131 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PresenterWidget.h"
+#include "Presentation.h"
+#include <AK/Format.h>
+#include <LibFileSystemAccessClient/Client.h>
+#include <LibGUI/Action.h>
+#include <LibGUI/Event.h>
+#include <LibGUI/Icon.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/Painter.h>
+#include <LibGUI/Window.h>
+#include <LibGfx/Forward.h>
+#include <LibGfx/Orientation.h>
+
+PresenterWidget::PresenterWidget()
+{
+    set_min_size(100, 100);
+}
+
+ErrorOr<void> PresenterWidget::initialize_menubar()
+{
+    auto* window = this->window();
+    // Set up the menu bar.
+    auto& file_menu = window->add_menu("&File");
+    auto open_action = GUI::CommonActions::make_open_action([this](auto&) {
+        auto response = FileSystemAccessClient::Client::the().try_open_file(this->window());
+        if (response.is_error())
+            return;
+        this->set_file(response.value()->filename());
+    });
+    auto about_action = GUI::CommonActions::make_about_action("Presenter", GUI::Icon::default_icon("app-display-settings"sv));
+
+    TRY(file_menu.try_add_action(open_action));
+    TRY(file_menu.try_add_action(about_action));
+
+    auto& presentation_menu = window->add_menu("&Presentation");
+    auto next_slide_action = GUI::Action::create("&Next", { KeyCode::Key_Right }, [this](auto&) {
+        if (m_current_presentation) {
+            m_current_presentation->next_frame();
+            outln("Switched forward to slide {} frame {}", m_current_presentation->current_slide_number(), m_current_presentation->current_frame_in_slide_number());
+            update();
+        }
+    });
+    auto previous_slide_action = GUI::Action::create("&Previous", { KeyCode::Key_Left }, [this](auto&) {
+        if (m_current_presentation) {
+            m_current_presentation->previous_frame();
+            outln("Switched backward to slide {} frame {}", m_current_presentation->current_slide_number(), m_current_presentation->current_frame_in_slide_number());
+            update();
+        }
+    });
+    TRY(presentation_menu.try_add_action(next_slide_action));
+    TRY(presentation_menu.try_add_action(previous_slide_action));
+    m_next_slide_action = next_slide_action;
+    m_previous_slide_action = previous_slide_action;
+
+    TRY(presentation_menu.try_add_action(GUI::Action::create("&Full Screen", { KeyModifier::Mod_Shift, KeyCode::Key_F5 }, { KeyCode::Key_F11 }, [this](auto&) {
+        this->window()->set_fullscreen(true);
+    })));
+    TRY(presentation_menu.try_add_action(GUI::Action::create("Present From First &Slide", { KeyCode::Key_F5 }, [this](auto&) {
+        if (m_current_presentation)
+            m_current_presentation->go_to_first_slide();
+        this->window()->set_fullscreen(true);
+    })));
+
+    return {};
+}
+
+void PresenterWidget::set_file(StringView file_name)
+{
+    auto presentation = Presentation::load_from_file(file_name, *window());
+    if (presentation.is_error()) {
+
```

Comment:
Empty line

Reply:
Hey, that's a bug source! This should create a message box.